src/Circulator.sol

```diff
@@ -187,16 +185,59 @@ contract Circulator is ICirculator, FeeOperator, Pausable, EIP712, Nonces {
             revert InvalidDelegateSignature();
         }
 
-        // Burn the token in TokenMessenger
-        nonce =
-            tokenMessenger.depositForBurn(burnAmt, delegateData.destinationDomain, delegateData.recipient, circleAsset);
+        nonce = _circulate(
+            burnAmount,
+            delegateData.outputAmount,
+            delegateData.recipient,
+            delegateData.destinationDomain,
+            delegateData.fillDeadline,
+            _type
+        );
 
         // Emit an event
         emit Circulate(
-            permitData.sender, delegateData.recipient, delegateData.destinationDomain, burnAmt, fee, nonce, msg.sender
+            permitData.sender,
+            delegateData.recipient,
+            delegateData.destinationDomain,
+            burnAmount,
+            fee,
+            nonce,
+            msg.sender,
+            _type
         );
     }
 
+    function _circulate(
+        uint256 _burnAmount,
+        uint256 _outputAmount,
+        address _recipient,
+        uint32 _destinationDomain,
+        uint32 _fillDeadline,
+        CirculateType _type
+    ) internal returns (uint64 nonce) {
+        if (_type == CirculateType.Cctp) {
```

Comment:
with `outputAmount` available 
we should make sure that burnAmount here is always >= outputAmount here
(so users can use `outputAmount` to make sure we don't charge a high fee)